metis-media-service/src/main/java/eu/europeana/metis/mediaprocessing/RdfDeserializerImpl.java

```diff
@@ -61,8 +145,78 @@
   @Override
   public List<String> getResourceEntriesForLinkChecking(InputStream inputStream)
       throws RdfDeserializationException {
-    return new ArrayList<>(
-        getResourceEntries(inputStream, UrlType.URL_TYPES_FOR_LINK_CHECKING).keySet());
+    return new ArrayList<>(getResourceEntries(deserializeToDocument(inputStream),
+            UrlType.URL_TYPES_FOR_LINK_CHECKING).keySet());
+  }
+
+  private static List<RdfResourceEntry> convertToResourceEntries(
+          Map<String, Set<UrlType>> urlWithTypes) {
+    return urlWithTypes.entrySet().stream().map(RdfDeserializerImpl::convertToResourceEntry)
+            .collect(Collectors.toList());
+  }
+
+  private static RdfResourceEntry convertToResourceEntry(Map.Entry<String, Set<UrlType>> entry) {
+    return new RdfResourceEntry(entry.getKey(), entry.getValue());
+  }
+
+  Map<String, Set<UrlType>> getResourceEntries(Document document,
+          Set<UrlType> allowedUrlTypes) throws RdfDeserializationException {
+    final Map<String, Set<UrlType>> urls = new HashMap<>();
+    for (UrlType type : allowedUrlTypes) {
+      final Set<String> urlsForType = getUrls(document, type);
+      for (String url : urlsForType) {
+        urls.computeIfAbsent(url, k -> new HashSet<>()).add(type);
+      }
+    }
+    return urls;
+  }
+
+  private Set<String> getUrls(Document document, UrlType type) throws RdfDeserializationException {
+
+    // Determine the right expression to apply.
+    final XPathExpression expression;
```

Comment:
I understand that the XPathExpression field of the class are usually all used right?
But in the case of thumbnail only one of those is used..
I think we can gain a lot of performance here if we only create one of those instead of 4.

Reply:
I'll do some on-demand evaluation that should solve this issue.

Reply:
ok